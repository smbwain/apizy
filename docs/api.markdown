---
layout: page
title: API Reference
permalink: /api/
nav_order: 3
---

> TODO

# API
{: .no_toc }

- TOC
{:toc}

## createApi

Api instance is needed to declare methods and types, and then expose it to a client.
Create API with _createApi_ function.

```ts
import {createApi} from 'apizy';

const api = createApi<Context = {}>();
```

Context can be useful if you want to share some things between handlers which will process every request.

E.g.

- authorization (@TODO: add example)
- access to raw http server (express) request object (@TODO: add example)
- db connection attached to client
- implementing cache on request level

## expose api

## Context

## Methods

```ts
createMethod<Output>(
    name: string,
    input: null,
    output: OutputType<Output>,
    handler: (input: void, context: Context) => Promise<Output>,
): void;

createMethod<Input, Output>(
  name: string,
  input: InputType<Input>,
  output: OutputType<Output>,
  handler: (input: Input, context: Context) => Promise<Output>,
): void;
```

## Entities

## Alias

## Custom types
