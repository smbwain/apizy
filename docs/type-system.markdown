---
layout: page
title: Type system
permalink: /type-system/
nav_order: 3
---

# type-system
{: .no_toc }

apizy type-system is used to describe types of your API data structures. Input and output.

- TOC
{:toc}

# In short, how does it work

An _apizy type-system_ is based on special _helpers_: javaScript functions that define types.

It allows _apizy_ to extract appropriate types from your definitions on compilation stage, as well as do runtime checks of entire input and output when running server and also generate typescript definitions for SDK.

E.g.

```ts
import {createApi, object, float, arrayOf, optional, string, nullable, uuid, boolean} from 'apizy';
const api = createApi();

api.createEntity(
    'methodName',
    object({
        a: float(),
        b: optional(string()),
        c: nullable(arrayOf(object({
            id: uuid(),
        }))),
    }),
    boolean(),
    async (input) => {
        // input has type {
        //      a: number;
        //      b?: string;
        //      c: Array<{
        //          id: string;
        //      }> | null;
        // }
    }
);
```

# Helpers

## Primitive types

### string

```ts
string()
```

### int

```ts
int()
```

### float

```ts
float()
```

### boolean

```ts
boolean()
```

### uuid

```ts
uuid()
```

### oneOf

For fields that must match one of several values, use the `oneOf` helper.

```ts
const userRole = oneOf(['admin', 'editor', 'subscriber']);

const userShape = object({
    id: uuid(),
    role: userRole,
});
```

This ensures the `role` field only accepts one of the predefined values.

## Structures

### object

Defines an object that uses multiple types for its fields,

```ts
object({
    id: uuid(),
    email: string(),
    privateData: object({
        firstName: string(),
        lastName: string(),
    }),
});
```

### arrayOf

```ts
const tagsShape = arrayOf(string());

const blogPostShape = object({
    title: string(),
    tags: tagsShape,
});
```

This ensures `tags` is an array of strings.

## Modifiers

### nullable

When a field can be `null`, use the `nullable` helper.

```ts
const userShape = object({
    id: uuid(),
    phoneNumber: nullable(string()),
});
```

`phoneNumber` can either be a string or `null`.

### optional

To allow a field to be optional (can be omitted in the input), use the `optional` helper.

```ts
const userShape = object({
    id: uuid(),
    nickname: optional(string()),
});
```

Here `nickname` may or may not be included in the object.

### defaultValue

The same as [optional](#optional), but allows to set default value.

```ts
const userShape = object({
    role: defaultValue('user', string()),
});
```

Here if the `role` field is not provided, it defaults to `'user'`.

## Relations

### extend

This helper allows you to add parts of output, which should not be loaded by default.
But it can be loaded (extended) only if a client app requested that.

You can do that, because of performance reason, or size of the object.

**Example**

```ts
import {createApi, object, float, arrayOf, optional, string, nullable, uuid, boolean} from 'apizy';
const api = createApi();

api.createEntity(
    'users.getById',
    string(),
    object({
        id: uuid(),
        name: string(),
        topArticles: extend(arrayOf(object({  // <-- extend helpers here
            id: uuid(),
            text: string(),
        }))),
    }),
    async (input) => {
        return {
            id: '...',
            name: '...',
            topArticles: async () => {
                // this function will be called to load articles,
                // only when user request this extension
                return [ {
                    id: '...',
                    text: '...',
                } /* ,...*/ ];
            },
        };
    }
);
```

On client side:
```ts
sdk.users.getById('...')
// Returns: Promise<{
//     id: string;
//     name: string;
// }>
```

```ts
sdk.users.getById('...', {
    extend: {
        topArticles: {},
    },
});
// Returns: Promise<{
//     id: string;
//     name: string;
//     topArticles: {
//         id: string;
//         text: string;
//     },
// }>
```

Extends can be nested.

### relation

TODO

> Check out [Entities](./api#entities) first.