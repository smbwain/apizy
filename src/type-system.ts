import * as check from 'checkeasy';
import {mapObject} from './utils';
import {TypeDescription} from './api-description';

export type PromiseOrValue<T> = Promise<T> | T;

export interface InputType<T> {
    extract: check.Validator<T>;
    desc: TypeDescription;
}

export interface OutputType<T> {
    resolve: (v: T, path: string, extendQuery: any, ctx: any) => PromiseOrValue<any>;
    checkExtendedQuery?: check.Validator<any>;
    desc: TypeDescription;
}

export interface Type<T> extends InputType<T>, OutputType<T> {}

export function isTypeInput<T>(t: InputType<T> | OutputType<T>): t is InputType<T> {
    return !!(t as any).extract;
}
export function isTypeOutput<T>(t: InputType<T> | OutputType<T>): t is OutputType<T> {
    return !!(t as any).resolve;
}

export function extend<T>(type: OutputType<T>): OutputType<() => PromiseOrValue<T>> {
    return {
        desc: {
            ...type.desc,
            outputExtendable: true,
        },
        resolve: async (v, path, extendQuery, ctx): Promise<any> => {
            if (!extendQuery) {
                return undefined;
            }
            return type.resolve(
                await v(),
                path,
                extendQuery,
                ctx,
            );
        },
        checkExtendedQuery: type.checkExtendedQuery || check.optional(check.object({})),
    };
}

export const string = (options?: Parameters<typeof check.string>[0]): Type<string> => ({
    extract: check.string(options),
    resolve: check.string(options),
    desc: {
        type: {
            ts: 'string',
        },
    },
});

export const uuid = (): Type<string> => ({
    extract: check.UUID(),
    resolve: check.UUID(),
    desc: {
        type: {
            ts: 'string',
        },
    },
});

export const int = (options?: Parameters<typeof check.int>[0]): Type<number> => ({
    extract: check.int(options),
    resolve: check.int(options),
    desc: {
        type: {
            ts: 'number',
        },
    },
});

export const float = (options?: Parameters<typeof check.float>[0]): Type<number> => ({
    extract: check.float(options),
    resolve: check.float(options),
    desc: {
        type: {
            ts: 'number',
        },
    },
});

export const boolean = (): Type<boolean> => ({
    extract: check.boolean(),
    resolve: check.boolean(),
    desc: {
        type: {
            ts: 'boolean',
        },
    },
});

// export function json<T>(scheme: check.Validator<T>): Type<T> {
//     return {
//         extract: (v, path) => {
//             let parsed;
//             try {
//                 parsed = JSON.parse(v);
//             } catch(err) {
//                 throw new check.ValidationError(`Not valid JSON not valid for [${path}]`);
//             }
//             return scheme(parsed, path);
//         },
//         resolve: (v) => JSON.stringify(v),
//         tsExtract: 'string',
//         tsResolve: 'string',
//     };
// }

export function object<O extends {[k: string]: Type<any>}>(description: O): Type<{
    [K in keyof O]: O[K] extends Type<infer T> ? T : never;
}>;
export function object<O extends {[k: string]: InputType<any>}>(description: O): InputType<{
    [K in keyof O]: O[K] extends InputType<infer T> ? T : never;
}>;
export function object<O extends {[k: string]: OutputType<any>}>(description: O): OutputType<{
    [K in keyof O]: O[K] extends OutputType<infer T> ? T : never;
}>;
export function object(description: any) {
    const keys = Object.keys(description);
    const res: any = {
        desc: {
            type: {
                objectOf: mapObject(description, (val: any) => val.desc),
            },
        },
    };
    if (keys.every(key => isTypeInput(description[key]))) {
        const d = description as Record<string, InputType<any>>;
        (res as InputType<any>).extract = check.object(mapObject(d, (val: any) => val.extract));
    }
    if (keys.every(key => isTypeOutput(description[key]))) {
        const d = description as Record<string, OutputType<any>>;
        (res as OutputType<any>).resolve = async (v, path, extendQuery, ctx) => {
            const res: any = {};
            await Promise.all(keys.map(async (key) => {
                res[key] = await d[key].resolve(v[key], `${path}.${key}`, extendQuery?.[key], ctx);
            }));
            return res;
        };

        const subValidators: Record<string, check.Validator<any>> = {};
        for (const key of keys) {
            if (d[key].checkExtendedQuery) {
                subValidators[key] = d[key].checkExtendedQuery!;
            }
        }

        (res as OutputType<any>).checkExtendedQuery = Object.keys(subValidators).length ? check.optional(check.object(subValidators)) : undefined;
    }
    return res;
}

export function arrayOf<T>(t: Type<T>, options?: Parameters<typeof check.arrayOf>[1]): Type<T[]>;
export function arrayOf<T>(t: InputType<T>, options?: Parameters<typeof check.arrayOf>[1]): InputType<T[]>;
export function arrayOf<T>(t: OutputType<T>, options?: Parameters<typeof check.arrayOf>[1]): OutputType<T[]>;
export function arrayOf(t: Type<any> | InputType<any> | OutputType<any>, options?: Parameters<typeof check.arrayOf>[1]) {
    const res: any = {
        desc: {
            type: {
                arrayOf: t.desc,
            },
        },
    };
    if (isTypeInput(t)) {
        (res as InputType<any>).extract = check.arrayOf(t.extract, options);
    }
    if (isTypeOutput(t)) {
        (res as OutputType<any>).resolve = async (v, path, extendedQuery, ctx) => {
            if (t.desc.outputExtendable && !extendedQuery) {
                return undefined;
            }
            return Promise.all((v as any[]).map((item, index) => t.resolve(item, `${path}[${index}]`, extendedQuery, ctx)))
        };
        (res as OutputType<any>).checkExtendedQuery = t.checkExtendedQuery;
    }
    return res;
}

export function nullable<T>(t: Type<T>): Type<T | null>;
export function nullable<T>(t: InputType<T>): InputType<T | null>;
export function nullable<T>(t: OutputType<T>): OutputType<T | null>;
export function nullable(t: Type<any> | InputType<any> | OutputType<any>) {
    const res: any = {
        desc: {
            ...t.desc,
            nullable: true,
        },
    };
    if (isTypeInput(t)) {
        (res as InputType<any>).extract = check.nullable(t.extract);
    }
    if (isTypeOutput(t)) {
        (res as OutputType<any>).resolve = async (v, path, extendedQuery, ctx) => {
            if (t.desc.outputExtendable && !extendedQuery) {
                return undefined;
            }
            if (v === null) {
                return null;
            }
            return t.resolve(v, path, extendedQuery, ctx);
        };
        (res as OutputType<any>).checkExtendedQuery = t.checkExtendedQuery;
    }
    return res;
}

export const optional = <T>(t: InputType<T>): InputType<T | undefined> => ({
    desc: {
        ...t.desc,
        inputOptional: true,
    },
    extract: check.optional(t.extract),
});

export const defaultValue = <T>(val: T, t: InputType<T>): InputType<T> => ({
    desc: {
        ...t.desc,
        inputOptional: true,
    },
    extract: check.defaultValue(val, t.extract),
});

export const oneOf = <T>(values: ReadonlyArray<T>): Type<T> => ({
    desc: {
        type: {
            ts: `${values.map(val => JSON.stringify(val)).join(' | ')}`,
        }
    },
    extract: check.oneOf(values),
    resolve: check.oneOf(values),
});

export const any = (): Type<any> => ({
    desc: {
        type: {
            ts: 'any',
        },
    },
    extract: v => v,
    resolve: v => v,
});

export const exact = <T>(value: T): Type<T> => ({
    desc: {
        type: {
            ts: JSON.stringify(value),
        },
    },
    extract: check.exact(value),
    resolve: check.exact(value),
});

export function relation<T>(
    type: OutputType<T>,
    load: (id: string, ctx: any) => Promise<T | null>,
    options: {
        nullable: true;
        auto?: boolean;
    },
): OutputType<string | null>;
export function relation<T>(
    type: OutputType<T>,
    load: (id: string, ctx: any) => Promise<T>,
    options?: {
        nullable?: false;
        auto?: boolean;
    },
): OutputType<string>;
export function relation<T>(
    type: OutputType<T>,
    load: (id: string, ctx: any) => Promise<T | null>,
    {nullable, auto = false}: {
        nullable?: boolean;
        auto?: boolean;
    } = {},
): OutputType<string> {
    return {
        desc: {
            ...type.desc,
            outputExtendable: !auto || undefined,
            nullable,
        },
        resolve: async (id, path, extendQuery, ctx) => {
            if (!auto && !extendQuery) {
                return undefined;
            }
            if (id === null) {
                if (nullable) {
                    return null;
                } else {
                    throw new Error(`Nullable is not allowed but null is passed as id [${path}]`);
                }
            }
            const res = await load(id, ctx);
            if (res === null) {
                if (nullable) {
                    return null;
                } else {
                    throw new Error(`Nullable is not allowed but no record found [${path}]`);
                }
            }
            return type.resolve(
                res,
                path,
                extendQuery,
                ctx,
            );
        },
        checkExtendedQuery: type.checkExtendedQuery || check.optional(check.object({})),
    };
}

export function describe<T>(t: Type<T>, description: string): Type<T>;
export function describe<T>(t: InputType<T>, description: string): InputType<T>;
export function describe<T>(t: OutputType<T>, description: string): OutputType<T>;
export function describe<T>(t: InputType<T> | OutputType<T>, description: string): InputType<T> | OutputType<T> {
    return {
        ...t,
        desc: {
            ...t.desc,
            description,
        },
    };
}

// export function alternatives<Alts extends ReadonlyArray<Type<any>>>(alts: Alts): Type<Alts extends ReadonlyArray<Type<infer T>> ? T : never>;
// export function alternatives<Alts extends ReadonlyArray<InputType<any>>>(alts: Alts): InputType<Alts extends ReadonlyArray<InputType<infer T>> ? T : never>;
// export function alternatives<Alts extends ReadonlyArray<OutputType<any>>>(alts: Alts): OutputType<Alts extends ReadonlyArray<OutputType<infer T>> ? T : never>;
// export function alternatives(alts: any[]) {
//     const res: any = {};
//     if (alts.every(alt => alt.extract)) {
//         const r: InputType<any> = {
//             tsExtract: alts.map(alt => `(${alt.tsExtract})`).join(' | '),
//             extract: check.oneOf(alts.map(alt => alt.extract)),
//         };
//         Object.assign(res, r);
//     }
//     // if (alts.every(alt => alt.resolve)) {
//     //     const r: OutputType<any> = {
//     //         tsResolve: alts.map(alt => `(${alt.tsExtract})`).join(' | '),
//     //         resolve: async (v, path, q, ctx) => Promise.all((v as any[]).map(item => t.resolve(item, `${path}[${item}]`, q, ctx))),
//     //         includeType: t.includeType,
//     //         tsInclude: t.tsInclude,
//     //     };
//     //     Object.assign(res, r);
//     // }
//     return res;
// }