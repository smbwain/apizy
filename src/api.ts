import {exact, InputType, object, OutputType, PromiseOrValue, string, Type} from './type-system';
import * as check from 'checkeasy';
import {BadReqError, NotFoundError} from './errors';
import {ApiDescription} from './api-description';

const NOT_INITIALIZED_METHOD = () => {
    throw new Error('Not initialized');
}

export interface Entity<Data, Context> extends OutputType<Data> {
    addResolver<Attrs extends {[k: string]: OutputType<any>}>(
        types: Attrs,
        handler: (data: Data, ctx: Context) => {
            [K in keyof Attrs]: Attrs[K] extends OutputType<infer T> ? T : never;
        },
    ): void;
}

export interface ApiContext {}

export interface Method<Context> {
    inputType: InputType<any> | null;
    outputType: OutputType<any>;
    handler: (input: any, ctx: Context) => Promise<any>;
}

export interface Api<Context extends ApiContext> {
    createMethod<Output>(
        name: string,
        input: null,
        output: OutputType<Output>,
        handler: (input: void, context: Context) => Promise<Output>,
    ): void;
    createMethod<Input, Output>(
        name: string,
        input: InputType<Input>,
        output: OutputType<Output>,
        handler: (input: Input, context: Context) => Promise<Output>,
    ): void;

    createEntity<Data>(
        name: string,
    ): Entity<Data, Context>;

    createAlias<T>(name: string, t: Type<T>): Type<T>;
    createAlias<T>(name: string, t: InputType<T>): InputType<T>;
    createAlias<T>(name: string, t: OutputType<T>): OutputType<T>;

    createCustomType<T>(
        name: string,
        tsDefinition: string,
        extract: check.Validator<T>,
        resolve: (v: T, path: string, extendQuery: any, ctx: any) => PromiseOrValue<any>,
        checkExtendedQuery?: check.Validator<any>,
        description?: string,
    ): Type<T>;
    createCustomType<T>(
        name: string,
        tsDefinition: string,
        extract: check.Validator<T>,
        resolve?: undefined,
        checkExtendedQuery?: undefined,
        description?: string,
    ): InputType<T>;
    createCustomType<T>(
        name: string,
        tsDefinition: string,
        extract: undefined,
        resolve: (v: T, path: string, extendQuery: any, ctx: any) => PromiseOrValue<any>,
        checkExtendedQuery?: check.Validator<any>,
        description?: string,
    ): OutputType<T>;

    callMethod(name: string, input: any, extend: any, ctx: Context): Promise<any>;

    apiDescription: ApiDescription;
}

export function createApi<Context extends ApiContext = ApiContext>(): Api<Context> {
    const apiDescription: ApiDescription = {
        methods: {},
        types: {},
    };

    const methods: {
        [methodName: string]: Method<Context>;
    } = {};

    function createMethod(name: string, inputType: InputType<any> | null, outputType: OutputType<any>, handler: (input: any, context: Context) => Promise<any>, description?: string) {
        if (methods[name]) {
            throw new Error('Method is already defined');
        }
        methods[name] = {
            inputType,
            outputType,
            handler,
        };
        apiDescription.methods[name] = {
            input: inputType?.desc,
            output: outputType.desc,
            description,
        };
    }

    function createEntity<Data>(
        name: string,
    ): Entity<Data, Context> {
        if (apiDescription.types[name]) {
            throw new Error('Type is already defined');
        }
        apiDescription.types[name] = {
            type: {
                objectOf: {},
            },
        };
        let checkExtendedQuery: check.Validator<any> = NOT_INITIALIZED_METHOD;
        let resolve: OutputType<any>['resolve'] = NOT_INITIALIZED_METHOD;
        const entity: Entity<any, Context> = {
            desc: {
                type: {
                    alias: name,
                },
            },
            checkExtendedQuery: (v, path) => checkExtendedQuery(v, path),
            resolve: (v, path, extendQuery, ctx) => resolve(v, path, extendQuery, ctx),
            addResolver: (
                types: Record<string, OutputType<any>>,
                handler: (data: any, ctx: Context) => any,
            ) => {
                const type = object({
                    $type: exact(name),
                    ...types,
                });
                resolve = async (v, path, extendQuery, ctx) => {
                    return type.resolve(
                        {
                            $type: name,
                            ...await handler(v, ctx),
                        },
                        path,
                        extendQuery,
                        ctx,
                    );
                };
                checkExtendedQuery = type.checkExtendedQuery ?? check.optional(check.object({}));
                apiDescription.types[name] = type.desc;
            },
        };

        return entity;
    }

    const createAlias = <T>(name: string, type: Type<T> | InputType<T> | OutputType<T>): any => {
        if (apiDescription.types[name]) {
            throw new Error('Type is already defined');
        }
        apiDescription.types[name] = type.desc;

        return {
            ...type,
            desc: {
                type: {
                    alias: name,
                },
            },
        };
    };

    const createCustomType = <T>(
        name: string,
        tsDefinition: string,
        extract?: check.Validator<T>,
        resolve?: (v: T, path: string, extendQuery: any, ctx: any) => PromiseOrValue<any>,
        checkExtendedQuery?: check.Validator<any>,
        description?: string,
    ): any => {
        if (apiDescription.types[name]) {
            throw new Error('Type is already defined');
        }
        apiDescription.types[name] = {
            type: {
                ts: tsDefinition,
            },
            description,
        };
        return {
            desc: {
                type: {
                    alias: name,
                },
            },
            extract,
            resolve,
            checkExtendedQuery,
        };
    }

    async function callMethod(name: string, unsafeInput: any, unsafeExtend: any, ctx: Context): Promise<any> {
        const method = methods[name];
        if (!method) {
            throw new NotFoundError('No method found');
        }
        const [input, extend]: any = (() => {
            try {
                return [
                    method.inputType?.extract(unsafeInput, 'input'),
                    (method.outputType.checkExtendedQuery ?? check.optional(check.object({})))(unsafeExtend, 'extend'),
                ];
            } catch (err) {
                if (err instanceof check.ValidationError) {
                    throw new BadReqError(err.message);
                }
                throw err;
            }
        })();
        return method.outputType.resolve(
            await method.handler(input, ctx),
            'output',
            extend,
            ctx,
        );
    }

    return {
        createEntity,
        createMethod,
        createAlias,
        createCustomType,
        callMethod,
        apiDescription,
    };
}