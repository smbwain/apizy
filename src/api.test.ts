import {arrayOf, boolean, defaultValue, extend, int, nullable, object, optional, string, uuid} from './type-system';
import {createApi} from './api';

describe('api', () => {
    it('should handle extend queries for entities', async () => {
        const api = createApi();

        const $Entity = api.createEntity<string>('Test');

        const $DateTime = api.createCustomType<string>(
            'DateTime',
            'string',
            (v, path) => {
                return '';
            },
            (v) => {
                return '';
            },
            undefined,
            'DateTime string in ISO format',
        );

        const $Money = api.createCustomType<number>(
            'Money',
            '{currency: string, amount: number}',
            undefined,
            (v) => ({
                currency: 'PLN',
                amount: v,
            }),
        );

        $Entity.addResolver({
            id: uuid(),
            date: $DateTime,
            registrationEndDate: $DateTime,
            duration: nullable(int()),
            participantsLimit: nullable(int()),
            participantsCount: int(),
            cancelled: boolean(),
            name: nullable(string()),
            price: nullable($Money),
            reducedPrice: nullable($Money),
            seatsIOEventKey: nullable(string()),
            prices: extend(arrayOf(string())),
            // priceAddons: extend(arrayOf($PriceAddon)),
        }, (ld, ctx) => {
            return null as any;
        });

        // $Entity.addResolver({
        //     prop1: string(),
        //     prop2: extend(string()),
        // }, () => {
        //     return {
        //         prop1: 'p1',
        //         prop2: () => 'p2',
        //     }
        // });
        //
        // expect($Entity.desc).toMatchSnapshot();
        //
        // $Entity.checkExtendedQuery!({}, 'extend query');
        //
        // $Entity.checkExtendedQuery!({prop2: {}}, 'extend query');
        //
        // expect(() => {
        //     $Entity.checkExtendedQuery!({prop1: {}}, 'extend query');
        // }).toThrowError();

        api.createMethod(
            'listings.dates.list',
            object({
                listingId: optional(uuid()),
                date: optional(object({
                    // from: optional($DateTime),
                    // to: optional($DateTime),
                })),
                limit: defaultValue(100, int({min: 1, max: 500})),
                page: defaultValue(0, int({min: 0})),
            }),
            object({
                items: arrayOf($Entity),
            }),
            async (filter, ctx) => {
                return {
                    items: [],
                };
            },
        );

        const res = await api.callMethod('listings.dates.list', {}, {items: {prices: {}}}, {});

        expect(res).toMatchSnapshot();

        // --

        api.createMethod(
            'dumbFunction',
            object({}),
            boolean(),
            async (filter, ctx) => {
                return true;
            },
        );

        const res2 = await api.callMethod('dumbFunction', {}, {}, {});
        expect(res2).toEqual(true);

        // expect( type.checkExtendedQuery!({}, 'extend query') ).toMatchSnapshot();
        //
        // expect( type.checkExtendedQuery!({prop2: {}}, 'extend query') ).toMatchSnapshot();
        //
        // expect(() => {
        //     type.checkExtendedQuery!({prop1: {}}, 'extend query');
        // } ).toThrowError(ValidationError);
    });
});