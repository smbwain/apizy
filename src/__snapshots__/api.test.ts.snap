// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`api should handle extend queries for entities 1`] = `
{
  "items": [],
}
`;
